fix large cluster deploy, deploy etcd slow

liupeng0518 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md and developer guide https://git.k8s.io/community/contributors/devel/development.md
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
1. When deploy a large cluster, ansible `copy` the etcd cert requires base64 first and then decode, which is very slow.
2. These are all executed in memory, which consumes a lot of memory and cpu.
**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #6911

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

LuckySB commented:
what do you mean when copy from item to item on one host ?

this task populate etcd certs generated on etcd[0] to all other nodes

LuckySB commented:
/hold

liupeng0518 commented:
> what do you mean when copy from item to item on one host ?
> 
> this task populate etcd certs generated on etcd[0] to all other nodes

Sorry,I was wrong, let me check.

liupeng0518 commented:
> what do you mean when copy from item to item on one host ?
> 
> this task populate etcd certs generated on etcd[0] to all other nodes

Here is copy from the first etcd node to all other nodes
```
when:
    - inventory_hostname in groups['etcd']
    - sync_certs|default(false)
    - inventory_hostname != groups['etcd'][0]
```
test:
```
TASK [etcd : Gen_certs | run cert generation script] *********************************************************************************
changed: [node1 -> 10.7.12.181]
Wednesday 30 December 2020  14:33:17 +0000 (0:00:00.847)       0:00:36.894 **** 

TASK [etcd : Gen_certs | Write etcd master certs] ************************************************************************************
changed: [node2] => (item=/etc/ssl/etcd/ssl/ca.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/ca.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/ca-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/ca-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/admin-node1.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/admin-node1.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/admin-node1-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/admin-node1-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/member-node1.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/member-node1-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/admin-node2.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/member-node1.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/admin-node2-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/member-node1-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/member-node2.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/member-node2-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/admin-node2.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/admin-node3.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/admin-node2-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/admin-node3-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/member-node2.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/member-node3.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/member-node2-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/member-node3-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/admin-node3.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node1.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/admin-node3-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node1-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/member-node3.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node2.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/member-node3-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node2-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node1.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node3.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node1-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node3-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node2.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node4.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node2-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node4-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node3.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node5.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node3-key.pem)
changed: [node2] => (item=/etc/ssl/etcd/ssl/node-node5-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node4.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node4-key.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node5.pem)
changed: [node3] => (item=/etc/ssl/etcd/ssl/node-node5-key.pem)
Wednesday 30 December 2020  14:33:27 +0000 (0:00:10.248)       0:00:47.142 **** 
```

LuckySB commented:
> Here is copy from the first etcd node to all other nodes

why from "first etcd node" ????

you mean, what `src: "{{ item }}"` - file on first etcd node ???

floryut commented:
ok thanks @LuckySB as I was reading changes I though the same and was wondering!

liupeng0518 commented:
> > Here is copy from the first etcd node to all other nodes
> 
> why from "first etcd node" ????
> 
> copy - copy file from LOCAL machine to dest
> so, if you launch ansible on separate admin-console-box-CI-runner, it will not have these files
> 
> uou should get all files from etcd-0 to ansible-launch host, and then put it to another etcd nodes

Okay, let me check.

liupeng0518 commented:
> we could slurp only once, using some run_once / delegate_to / delegate_facts (and we would only have the certs once in ram)

For example, with 500 nodes, there are too many cert, and loading at one time is very slow. May take 10 hours.

liupeng0518 commented:
/close
